Add leds_gpio driver and modify blinky application

Mani-Sadhasivam opened the pull request:
This PR adds leds_gpio driver which is used to control the LED devices attached to the GPIO pins through LED API and also modifies the existing blinky application as a consumer.

The leds_gpio driver currently supports only `led_on` and `led_off` APIs and there are plans to extend this driver to use timers for software blinking implementation in future.

The driver requires no change in the board configuration at all. It has been designed to reuse the `gpio-leds` configuration for compatibility and supports only two LED devices for now. Both driver and sample application are validated on 96Boards Carbon board.

Signed-off-by: Manivannan Sadhasivam <manivannan.sadhasivam@linaro.org>

Mani-Sadhasivam committed:
led: Add driver for controlling LEDs attached to GPIOs

Add driver for controlling the LED devices attached to GPIO pins
which are declared in devicetree under leds node. The driver takes
care of configuring the pins to output state before using it.

Signed-off-by: Manivannan Sadhasivam <manivannan.sadhasivam@linaro.org>

Mani-Sadhasivam committed:
samples: basic: blinky: Convert application to use leds_gpio driver

leds_gpio driver is used to control the LED devices attached to GPIO
pins onboard. Hence, convert the blinky application to use this driver
instead of toggling the GPIO pin manually.

Signed-off-by: Manivannan Sadhasivam <manivannan.sadhasivam@linaro.org>

Mani-Sadhasivam commented:
@erwango This `leds_gpio` requires the boards to have `leds` node for proper working. Without this node, we will run into `LED0_GPIO_CONTROLLER` undefined error.

Currently we have the following in `blinky` sample which takes care of this scenario:

```
/* Change this if you have an LED connected to a custom port */
#ifndef LED0_GPIO_CONTROLLER
#define LED0_GPIO_CONTROLLER    LED0_GPIO_PORT
#endif
```

Any idea of how we can handle this in driver? Moving the above defines to driver itself?

erwango commented on drivers/led/Kconfig.gpio:
```diff
@@ -0,0 +1,47 @@
+
+if LED0_GPIO
+
+config LED0_GPIO_NAME
```
Can't we use the 'label' to get the instance name?

Mani-Sadhasivam replied:
okay.

erwango commented on drivers/led/Kconfig.gpio:
```diff
@@ -0,0 +1,47 @@
+
+# ---------- LED 0 ----------
+
+config LED0_GPIO
```
I'm finding 'LED0_GPIO' a bit confusing, as I read it 'GPIO of LED0'.
Would LED_GPIO_0 be a bit less confusing?

Mani-Sadhasivam replied:
Actually, the naming convention aligns with the DTS generated header (LEDx_GPIO). But anyway, `LED_GPIO_x` seems to be good.

erwango replied:
@Mani-Sadhasivam , sorry but I still find it a bit confusing actually. What about GPIO_LED ?
@MaureenHelm , any opinion on that?

erwango commented on drivers/led/leds_gpio.c:
```diff
@@ -0,0 +1,114 @@
+	struct led_data dev_data;
+};
+
+static inline int led_gpio_on(struct device *dev, u32_t led)
```
Is 'led' argument actually required ?
We should be able to retrieve it from *dev. 
And I think this is the interest of the API, so user doesn't care of the PIN

Mani-Sadhasivam replied:
`led` argument is there only for the purpose of LED API and yes, we can retrieve it from *dev. Anyway, passing the `led` argument gives us the possibility of checking the correct pin in led_on/led_off as below:

```c
        /* Make sure that the LED is valid */
        if (led != cfg->gpio_pin)
                return -EINVAL;
```
Will include the above check.

erwango replied:
@Mani-Sadhasivam , 
> Anyway, passing the led argument gives us the possibility of checking the correct pin in led_on/led_off

In the other hand not having 'led' as arguments makes this kind of issue impossible.
I still prefer this is removed from the API.

MaureenHelm replied:
I agree the argument should stay and the driver should check it.

erwango replied:
Can you check following PR to see if it wouldn't provide similar service and spare the argument?
https://github.com/zephyrproject-rtos/zephyr/pull/12056

pabigot replied:
I don't see how #12056 relates.  I also agree that since this is being done as an implementation for an API that is designed to support multiple LEDs the argument must stay.  However, see review summary.

erwango reviewed (dismissed):
Some comments on otherwise good stuf

dbkinder commented on samples/basic/blinky/README.rst:
```diff
@@ -6,9 +6,9 @@ Blinky Application
-The Blinky example shows how to configure GPIO pins as outputs which can also be
-used to drive LEDs on the hardware usually delivered as "User LEDs" on many of
-the supported boards in Zephyr.
+The Blinky example shows how to drive LED devices attached to GPIO
```
Change ``drive LED devices`` to ``configure LED devices``. To me, the word ``drive`` has hardware implications (voltage, resistors, and such), whereas the previous word ``configure`` feels better here since it's a software configuration we're discussing, right?

Mani-Sadhasivam replied:
Okay

Mani-Sadhasivam replied:
Actually, in this PR I have moved the configuration part inside the LED GPIO driver. So we don't configure pins in this sample anymore.

Mani-Sadhasivam commented:
@erwango I have updated the PR incorporating your suggestions. Still, we have to figure out how to solve issues with boards without `leds` node.

dbkinder reviewed (approved):
OK

MaureenHelm commented on drivers/led/leds_gpio.c:
```diff
@@ -0,0 +1,124 @@
+	if (led != cfg->gpio_pin)
+		return -EINVAL;
+
+	gpio_pin_write(data->gpio, led, LED_ON);
```
Many NXP boards invert the polarity, meaning you have to write a zero to the gpio to turn on the LED.

Mani-Sadhasivam replied:
Okay. Then the best way to address is to have a `invert-polarity` property on the `leds` node. What is your opinion?

Mani-Sadhasivam replied:
@MaureenHelm Just a ping over my query.

MaureenHelm replied:
We've already got information about the polarity in dts, can you use that?
https://github.com/zephyrproject-rtos/zephyr/blob/c1f54cc5fdec8e5180431aae151e5f4299944367/boards/arm/frdm_k64f/frdm_k64f.dts#L50-L51

Mani-Sadhasivam replied:
Hmm, makes sense. Thanks for the pointer!

MaureenHelm commented on drivers/led/leds_gpio.c:
```diff
@@ -0,0 +1,142 @@
+	if (cfg->gpio_polarity)
+		gpio_pin_write(data->gpio, led, LED_ON);
+	else
+		gpio_pin_write(data->gpio, led, LED_OFF);
```
How about replacing the if/else with:
```c
gpio_pin_write(data->gpio, led, (cfg->gpio_polarity & GPIO_INT_ACTIVE_HIGH) != 0);
```

Mani-Sadhasivam replied:
yeah, but I just wanted to make it explicit for the user here. Anyway, I can modify this.

MaureenHelm commented on drivers/led/leds_gpio.c:
```diff
@@ -0,0 +1,142 @@
+
+	/* Make sure that the LED is valid */
+	if (led != cfg->gpio_pin)
+		return -EINVAL;
```
Wrap this line in braces. There are a few other places in the driver that need braces too

Mani-Sadhasivam replied:
Okay

MaureenHelm commented on drivers/led/leds_gpio.c:
```diff
@@ -0,0 +1,142 @@
+	if (cfg->gpio_polarity)
+		gpio_pin_write(data->gpio, led, LED_OFF);
+	else
+		gpio_pin_write(data->gpio, led, LED_ON);
```
Similar to above, try this instead of the if/else:
```c
gpio_pin_write(data->gpio, led, (cfg->gpio_polarity & GPIO_INT_ACTIVE_HIGH) == 0);
```

Mani-Sadhasivam replied:
Okay

MaureenHelm reviewed (changes requested):
Please fix the CI failures.

The merge window is currently closed to new features, so we need to wait to merge this PR until after the 1.13 release is complete.

MaureenHelm commented:
> We need to add a few more for FLAGS, LABEL etc... So I just asked before that whether we need to have these #ifdefs or any other better solution available?

The better solution is to add led nodes to the board dts, otherwise we need the defines.

MaureenHelm commented:
@erwango Please have another look

Mani-Sadhasivam commented:
@erwango Can you please take a look?

Mani-Sadhasivam commented:
Build failed due to lack of onboard LED support in `em_starterkit` boards. Do I need to add the onboard led support or is there any work going on?

erwango commented:
@Mani-Sadhasivam 

> Build failed due to lack of onboard LED support in em_starterkit boards. Do I need to add the onboard led support or is there any work going on?

I think it is ok if you do.

Mani-Sadhasivam committed:
led: lp3943: Make default Kconfig definitions depend on I2C

Since LP3943 is an I2C device, relevant Kconfig symbols should only be
used when I2C is enabled.

Signed-off-by: Manivannan Sadhasivam <manivannan.sadhasivam@linaro.org>

Mani-Sadhasivam commented:
@erwango Mind taking a look at the updated rev?

erwango commented:
@Mani-Sadhasivam , 

> @erwango Mind taking a look at the updated rev?

Can you sum up the changeson this update? This has been a while ... :-)

Mani-Sadhasivam commented:
@erwango Nothing much really... There was a build failure when you last reviewed but that got fixed when LED node was added to `em_starterkit` board.

Anyway, here is the changeset:

* Rebased to master
* Fixed the LP3943 Kconfig to enable the default options only when CONFIG_I2C is set. Otherwise,
it will give errors as `!HAS_DTS_I2C` will become true.

pabigot reviewed (changes requested):
I think a GPIO LED driver is very much needed; all existing uses of GPIO LEDs ignore the active level which is annoying when working on cross-platform applications.

However, I also **really** don't like having a distinct driver instance for each GPIO LED on the board: the boards I use most have 4 to 6 LEDs.  IMO this should be reworked so that each `led-gpio` instance supports multiple LEDs, just like all the other implementations of the generic LED API.  I think it would be fair to require that the LEDs be located on the same GPIO device and have the same active level so that information need not be replicated; only the pin selector for each LED would be needed.

This does get a little tricky to do generically until device-tree support is more robust, though if there's agreement this is worth pursuing I do have some ideas.  But honestly I'd rather not have this driver available than have it merged with this design, which is why I'm currently nacking the PR.  If I'm the only one who feels this way I'll remove the nack.

The only board I'm immediately aware of that has GPIO LEDs on different gpio devices are the Particle boards proposed in #12195; for there there would have to be two devices, but the LEDs on each gpio device have different roles so that's not a bad thing.

Mani-Sadhasivam commented:
Closing at it got superseded.
